File: quant-black-box/backend/app/market.py
"""London Strategic Edge market data client.

Wraps the lse-data Python SDK to fetch candles, quotes, options chains,
catalog entries, and macro series. Authenticates with the LSE_API_KEY
environment variable.
"""

import os
from datetime import datetime, timezone


def _get_client():
    """Lazy-import and instantiate the LSE client."""
    from lse import LSE

    api_key = os.environ.get("LSE_API_KEY")
    if not api_key:
        raise RuntimeError(
            "LSE_API_KEY not set. Add it to backend/.env or your environment. "
            "Get a key at https://londonstrategicedge.com/data"
        )
    return LSE(api_key=api_key)


def get_candles(
    symbol: str,
    timeframe: str = "1d",
    start: str | None = None,
    end: str | None = None,
    limit: int = 500,
) -> list[dict]:
    """Fetch OHLCV candles from the LSE vault."""
    client = _get_client()
    rows = client.candles(
        symbol,
        timeframe,
        start=start,
        end=end,
        limit=limit,
        order="asc",
    )
    return rows


def get_quote(symbol: str) -> dict:
    """Get the latest price for a symbol.

    Fetches the most recent candle and extracts the close price.
    """
    client = _get_client()
    rows = client.candles(symbol, "1d", limit=1, order="desc")
    if not rows:
        raise ValueError(f"No data found for symbol: {symbol}")
    row = rows[-1]
    return {
        "symbol": symbol,
        "price": row.get("close", row.get("c", 0)),
        "bid": row.get("bid", row.get("b", 0)),
        "ask": row.get("ask", row.get("a", 0)),
        "volume": row.get("volume", row.get("v", 0)),
        "timestamp": row.get("timestamp", row.get("ts", row.get("datetime", ""))),
        "name": row.get("name", symbol),
    }


def get_options_chain(underlying: str) -> list[dict]:
    """Fetch the options chain for an underlying with IV and Greeks."""
    client = _get_client()
    rows = client.options(underlying)
    contracts = []
    for r in rows:
        contracts.append({
            "ticker": r.get("ticker", ""),
            "strike": r.get("strike", 0),
            "expiry": r.get("expiry", r.get("expiration", "")),
            "type": r.get("type", r.get("call_put", "")),
            "price": r.get("price", r.get("last", 0)),
            "iv": r.get("iv", r.get("implied_volatility", 0)),
            "delta": r.get("delta", 0),
            "gamma": r.get("gamma", 0),
            "theta": r.get("theta", 0),
            "vega": r.get("vega", 0),
            "rho": r.get("rho", 0),
            "volume": int(r.get("volume", 0)),
            "premium": r.get("premium", 0),
        })
    return contracts


def get_catalog(category: str | None = None) -> list[dict]:
    """List available instruments from the vault catalog."""
    client = _get_client()
    rows = client.catalog(category) if category else client.catalog()
    entries = []
    for r in rows:
        entries.append({
            "symbol": r.get("symbol", ""),
            "name": r.get("name", ""),
            "category": r.get("category", ""),
            "ticks": r.get("ticks", 0),
            "first": r.get("first"),
            "last": r.get("last"),
        })
    return entries


def get_series(
    symbol: str,
    start: str | None = None,
    end: str | None = None,
    limit: int = 500,
) -> list[dict]:
    """Fetch a macro series or bond yield time series."""
    client = _get_client()
    rows = client.series(symbol, start=start, end=end, limit=limit, order="asc")
    result = []
    for r in rows:
        date_val = r.get("date", r.get("datetime", r.get("ts", "")))
        value_val = r.get("value", r.get("v", 0))
        result.append({"date": str(date_val), "value": float(value_val)})
    return result

File: quant-black-box/server/src/market.ts
import { env } from './config.js';

const BASE = 'https://api.londonstrategicedge.com/vault';

async function lseFetch(path: string, params: Record<string, string> = {}): Promise<unknown> {
  const url = new URL(`${BASE}${path}`);
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== '') url.searchParams.set(k, v);
  }
  const res = await fetch(url, { headers: { 'x-api-key': env.LSE_API_KEY } });
  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new Error(`LSE API error ${res.status}: ${body}`);
  }
  return res.json();
}

export async function getQuote(symbol: string) {
  const rows = await lseFetch('/candles', { symbol, timeframe: '1d', limit: '1', order: 'desc' }) as Record<string, unknown>[];
  if (!rows || rows.length === 0) throw new Error(`No data for ${symbol}`);
  const row = rows[rows.length - 1]!;
  return {
    symbol,
    price: Number(row.close ?? row.c ?? 0),
    bid: Number(row.bid ?? row.b ?? 0),
    ask: Number(row.ask ?? row.a ?? 0),
    volume: Number(row.volume ?? row.v ?? 0),
    timestamp: String(row.timestamp ?? row.ts ?? row.datetime ?? ''),
    name: String(row.name ?? symbol),
  };
}

export async function getCandles(symbol: string, timeframe = '1d', start?: string, end?: string, limit = 500) {
  const rows = await lseFetch('/candles', { symbol, timeframe, start: start || '', end: end || '', limit: String(limit), order: 'asc' }) as Record<string, unknown>[];
  return rows.map((r) => ({
    timestamp: String(r.timestamp ?? r.ts ?? r.datetime ?? ''),
    open: Number(r.open ?? r.o ?? 0),
    high: Number(r.high ?? r.h ?? 0),
    low: Number(r.low ?? r.l ?? 0),
    close: Number(r.close ?? r.c ?? 0),
    volume: Number(r.volume ?? r.v ?? 0),
  }));
}

export async function getOptionsChain(underlying: string) {
  const rows = await lseFetch('/options', { underlying }) as Record<string, unknown>[];
  return rows.map((r) => ({
    ticker: String(r.ticker ?? ''),
    strike: Number(r.strike ?? 0),
    expiry: String(r.expiry ?? r.expiration ?? ''),
    type: String(r.type ?? r.call_put ?? ''),
    price: Number(r.price ?? r.last ?? 0),
    iv: Number(r.iv ?? r.implied_volatility ?? 0),
    delta: Number(r.delta ?? 0),
    gamma: Number(r.gamma ?? 0),
    theta: Number(r.theta ?? 0),
    vega: Number(r.vega ?? 0),
    rho: Number(r.rho ?? 0),
    volume: Number(r.volume ?? 0),
    premium: Number(r.premium ?? 0),
  }));
}

export async function getCatalog(category?: string) {
  const rows = await lseFetch('/catalog', category ? { category } : {}) as Record<string, unknown>[];
  return rows.map((r) => ({
    symbol: String(r.symbol ?? ''),
    name: String(r.name ?? ''),
    category: String(r.category ?? ''),
    ticks: Number(r.ticks ?? 0),
    first: r.first != null ? String(r.first) : null,
    last: r.last != null ? String(r.last) : null,
  }));
}

export async function getSeries(symbol: string, start?: string, end?: string, limit = 500) {
  const rows = await lseFetch('/series', { symbol, start: start || '', end: end || '', limit: String(limit), order: 'asc' }) as Record<string, unknown>[];
  return rows.map((r) => ({
    date: String(r.date ?? r.datetime ?? r.ts ?? ''),
    value: Number(r.value ?? r.v ?? 0),
  }));
}
